src/System.Management.Automation/engine/Attributes.cs

```diff
@@ -943,43 +970,100 @@ protected override void ValidateElement(object element)
                         Metadata.ValidateNotNullFailure);
             }
 
-            var o = element as PSObject;
-            if (o != null)
+            if(_rangeKind.HasValue)
             {
-                element = o.BaseObject;
+                // Using dynamic to handle different type of numbers such as
+                // int, double, decimal and float. 
+                dynamic val = element;
```

Comment:
>This might be too general - e.g. strings implement  IComparable , but strings aren't positive or negative.

In general, we cannot know if the type is comparable with zero in advance. Should we test this?